### ISSUE TITLE
Convert ACT model (.safetensor) from LeRobot to .vmfb for running on sl2619

### ISSUE THREAD

**Issue by vietnv** [synaptics-torq/torq-compiler#17]
[lower_to_sl2610.sh](https://github.com/user-attachments/files/24883643/lower_to_sl2610.sh)
[convert_model.py](https://github.com/user-attachments/files/24883642/convert_model.py)

Hi,
We have the pretrained model at .safetensor format and we have issue when using torq-compiler tool to convert it to .vmfb ( target device: sl2619)
Following are steps:
1. Convert .safetensor to .onnx: Success (using convert_model.py at attachment)
2. Convert .onnx to .mlir and lower to linagl backend successfully. But compile from .mlir to .vmfb: Failed ( torq-compiler crash at reshape() func) ( Using lower_to_sl2610.py at attachment) 
Here are failure logs:
`root@7ea2215dd03c:/home/vietnv/SL2610# ./lower_to_sl2610.sh 
Step 1: Importing ONNX to Torch MLIR...
Step 2: Lowering Torch Dialect to Linalg...
Step 3: Compiling to .vmfb...
Attempting compilation with 'torq' backend...
torq-compile: /workspace/code/third_party/iree/third_party/llvm-project/mlir/lib/IR/BuiltinAttributes.cpp:1237: DenseElementsAttr mlir::DenseElementsAttr::reshape(ShapedType): Assertion `newType.getNumElements() == curType.getNumElements() && "expected the same number of elements"' failed.
Please report issues to https://github.com/iree-org/iree/issues and include the crash backtrace.
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libIREECompiler.so 0x00007a7d015ade18 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 40
1  libIREECompiler.so 0x00007a7d015ac030 llvm::sys::RunSignalHandlers() + 80
2  libIREECompiler.so 0x00007a7d015ae4b8
3  libc.so.6          0x00007a7cfc3d7330
4  libc.so.6          0x00007a7cfc430b2c pthread_kill + 284
5  libc.so.6          0x00007a7cfc3d727e gsignal + 30
6  libc.so.6          0x00007a7cfc3ba8ff abort + 223
7  libc.so.6          0x00007a7cfc3ba81b
8  libc.so.6          0x00007a7cfc3cd517
9  libIREECompiler.so 0x00007a7d0160a8cf
10 libIREECompiler.so 0x00007a7d0527f517 mlir::tensor::ExpandShapeOp::fold(mlir::tensor::ExpandShapeOpGenericAdaptor<llvm::ArrayRef<mlir::Attribute>>) + 343
11 libIREECompiler.so 0x00007a7d05252d82
12 libIREECompiler.so 0x00007a7d052523b0
13 libIREECompiler.so 0x00007a7d01689b3d mlir::Operation::fold(llvm::ArrayRef<mlir::Attribute>, llvm::SmallVectorImpl<mlir::OpFoldResult>&) + 45
14 libIREECompiler.so 0x00007a7d01689ec3 mlir::Operation::fold(llvm::SmallVectorImpl<mlir::OpFoldResult>&) + 355
15 libIREECompiler.so 0x00007a7d050d5525
16 libIREECompiler.so 0x00007a7d050d2801 mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) + 1345
17 libIREECompiler.so 0x00007a7d022d0888
18 libIREECompiler.so 0x00007a7d0174d7cf mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 639
19 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
20 libIREECompiler.so 0x00007a7d0175315a
21 libIREECompiler.so 0x00007a7d0174f6ab mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 2427
22 libIREECompiler.so 0x00007a7d0174d961 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1041
23 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
24 libIREECompiler.so 0x00007a7d01752441
25 libIREECompiler.so 0x00007a7d01c30a95
26 libIREECompiler.so 0x00007a7d0174d7cf mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 639
27 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
28 libIREECompiler.so 0x00007a7d0175315a
29 libIREECompiler.so 0x00007a7d0174f6ab mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 2427
30 libIREECompiler.so 0x00007a7d0174d961 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1041
31 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
32 libIREECompiler.so 0x00007a7d01752441
33 libIREECompiler.so 0x00007a7d028c77fb
34 libIREECompiler.so 0x00007a7d0174d7cf mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 639
35 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
36 libIREECompiler.so 0x00007a7d0175315a
37 libIREECompiler.so 0x00007a7d0174f6ab mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 2427
38 libIREECompiler.so 0x00007a7d0174d961 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1041
39 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
40 libIREECompiler.so 0x00007a7d01752441
41 libIREECompiler.so 0x00007a7d028c82ba
42 libIREECompiler.so 0x00007a7d0174d7cf mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 639
43 libIREECompiler.so 0x00007a7d0174e097 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 311
44 libIREECompiler.so 0x00007a7d0175315a
45 libIREECompiler.so 0x00007a7d0175320e
46 libIREECompiler.so 0x00007a7d016ad2a8
47 libIREECompiler.so 0x00007a7d016ad207
48 libc.so.6          0x00007a7cfc433ed3
49 libIREECompiler.so 0x00007a7d016ad5df
50 libIREECompiler.so 0x00007a7d016ad665
51 libIREECompiler.so 0x00007a7d015695e9 llvm::StdThreadPool::processTasks(llvm::ThreadPoolTaskGroup*) + 985
52 libIREECompiler.so 0x00007a7d0156aafc
53 libc.so.6          0x00007a7cfc42eaa4
54 libc.so.6          0x00007a7cfc4bba64 __clone + 68
./lower_to_sl2610.sh: line 52:  5751 Aborted                 (core dumped) torq-compile "lerobot_linalg.mlir" -o $OUTPUT_VMFB
torq compilation failed. Trying with iree-compile...
./lower_to_sl2610.sh: line 52:  5757 Segmentation fault      (core dumped) /opt/release/tools/iree-compile "lerobot_linalg.mlir" --iree-hal-target-backends=torq -o $OUTPUT_VMFB 2> /dev/null
Compilation with torq failed. Attempting with CPU backend (no NPU optimization)...
SUCCESS: Generated CPU-only VMFB. The model logic is valid!
The crash is specific to Synaptics NPU optimization passes.
------------------------------------------------
Done! Generated: lerobot_act_sl2619.vmfb
You can now run this model on the SL2610 target.`

Per the log, if we use iree-compile dirrectly with target = llvm-cpu, the model can be compiled successfully to .vmfb
Could you help to fix this issue for optimize model to run on Torq NPU.

===========================================================
P/S: 
- We are using torq_compile via docker 
- Per private ticket case (Sauryadeep.Pal), we have use following script to check if the model contains DYNAMIC TENSORS.  it's all STATIC
`import onnx
def check_for_dynamic_shapes(model_path):
    model = onnx.load(model_path)
    graph = model.graph
    
    print(f"Checking model: {model_path}")
    print("-" * 30)
    

    for input in graph.input:
        name = input.name
        # Get dimensions
        dims = [d.dim_value if d.dim_value > 0 else "DYNAMIC" 
                for d in input.type.tensor_type.shape.dim]
        
        if "DYNAMIC" in dims:
            print(f"⚠️ INPUT '{name}' is DYNAMIC: {dims}")
        else:
            print(f"✅ INPUT '{name}' is Static: {dims}")

    for output in graph.output:
        name = output.name
        dims = [d.dim_value if d.dim_value > 0 else "DYNAMIC" 
                for d in output.type.tensor_type.shape.dim]
        
        if "DYNAMIC" in dims:
            print(f"⚠️ OUTPUT '{name}' is DYNAMIC: {dims}")
        else:
            print(f"✅ OUTPUT '{name}' is Static: {dims}")

check_for_dynamic_shapes("lerobot_model.onnx")

And the output:
`Checking model: lerobot_model.onnx
------------------------------
✅ INPUT 'image_side' is Static: [1, 3, 480, 640]
✅ INPUT 'state' is Static: [1, 6]
✅ OUTPUT 'action' is Static: [1, 100, 6]`

Here is the link of our .safetensor model: https://drive.google.com/file/d/1HoPjDXzaa6vmN58yYbqoqrzG7zvAewjm/view?usp=sharing
Best regards
Viet Nguyen.

**Comment by nchristo-synaptics** [comment]
Hi @vietnv -- could you provide some details regarding what runtime speed (fps) you are expecting out of this model on the SL2610? If there's any other information you think would be relevant let us know. Thanks!

**Comment by nchristo-synaptics** [comment]
@vietnv one more request -- do you have a collection input images on which the model was trained to validate the model on our end?

**Comment by vietnv** [comment]
@nchristo-synaptics 
Sorry for this late response.
Since we are evaluating whether we can run a converted ACT policy on SL2619 or not:
fps would not be important as of now, (was expected ~ 10fps with INT8 quantized). Any fps would be ok for us. Let's optimize it later when we have the policy work on the kit.
- The 5-second dataset is here: https://huggingface.co/datasets/quan1328/record-test
- The ACT policy trained from the dataset is here: https://huggingface.co/quan1328/so101_recordtest_torq_policy

One more question: Any updates on this ACT model  for the inference on board sl2619 work with SO-101 Arm ( le-robot framework), and how about SmoVLA model as well? Is there any inference models available  on sl2619 ( optimize for running on NPU) that we can try? 
Thanks

**Comment by nchristo-synaptics** [comment]
Thanks! We have the ACT model running on SL2610 at about 1 inference per 1.67seconds, as compared with ORT at 1inf/3.35s on the same SL2610. I'm actively working on this and will update you when I have more done.

**Comment by nchristo-synaptics** [comment]
Hi @vietnv, this dataset has no action signal to learn from. Across all 5 episodes and 414 frames, both observation.state and action are effectively constant (every frame is the same pose; action std \approx 0). The policy learned this perfectly and returns the same constant action for any input, including random noise (confirmed on the PyTorch policy, fp32 ONNX, and our pipeline).

The pipeline itself checks out numerically, but I can't measure int8 quantization accuracy when the output doesn't respond to inputs -- any quantization looks lossless. Could you share a dataset with real motion and a model trained on it? With those I can validate int8/bf16 accuracy properly.

**Comment by quanguyen** [comment]
Hi @nchristo-synaptics ,

Thanks for catching that our earlier dataset was effectively static — that flag saved us from "validating" a model that couldn't really be tested.

We'd like to confirm the ACT → Torq conversion and the `int8`/`bf16` quantization are sound before moving to hardware. The questions below are all on-paper checks — none require the robot arm.

Questions

1.  Where the model runs. Could you share which parts of the model run on the NPU versus the main CPU, and how many times execution switches between them? [op-placement report; CPU fallback; subgraph count]
2. Random part fixed. Is ACT's random "style" variable locked to its default in the deployed model, so the output is the same on every run? [latent z / VAE — deterministic inference path]
3. Accuracy after shrinking. On real-motion data, could you show how far the `int8` and `bf16` outputs drift from `fp32` — per joint, and across the full predicted motion, not just the first step? [per-joint MAE / L1]
4. Shrink method and fallback. Did you quantize after training or with retraining? And if int8 costs too much accuracy, could we keep the sensitive layers at higher precision? [PTQ vs QAT; per-channel; mixed precision]
5. Conversion check, before hardware. Could you share a short report showing the converted `fp32` model matches the original model's outputs closely, alongside the op-placement and accuracy numbers? We'd like to review the results on paper before moving to hardware.
6. A note on our data, and a stronger test. To be upfront: our current episodes use very simple, repetitive motion, so quantization may look near-lossless mainly because the task is easy — which wouldn't fully prove the pipeline. For a fairer test, could you also run it on a community model trained on richer, more varied motion?

- Our dataset: https://huggingface.co/datasets/quan1328/record-test1
- Our trained ACT model: https://huggingface.co/quan1328/so101_recordtest_torq_policy13
- Community model to test: https://huggingface.co/lerobot/act_aloha_sim_insertion_human

Suggested order: confirm the pipeline handles our model first, then use the community model to measure accuracy on real motion.

 7. Reproducing it ourselves. Could you share the scripts, settings, and tool/software versions, so our team can repeat the process on our own models later? [reproducible pipeline]

Thanks again — happy to share more detail or grant access on any of the above.

Regards,

Quan

**Comment by nchristo-synaptics** [comment]
Hi @quanguyen. Support is still ongoing; here's where we currently are. Before I got this information I was working on this trained model (https://huggingface.co/tatsuya-25/1camera_3) to answer my questions about accuracy. Here are the results relative to that model.

1. The end-to-end pipeline is in mixed precision and broken into 3 pieces: the Resnet CNN followed by the transformer split into two chunks. Host use is limited to orchestration and a single int8 -> bf16 cast at the boundary of the ResNet and the Transformer portions of the model. Runtime is currently at ~700 ms / inference.
2. Yes, exactly. There are no stochastic ops on the model at inference.
3. Our current best (mixed-precision) model: per-joint MAE ≤0.14% of joint range, 0.12% overall (absolute: ≤0.005 per joint, 0.0043 overall), worst absolute error |0.197|, overall relL2 0.752%. Drift is flat across all 100 predicted steps — it does not grow. See (5) for the full per-element data and a summary, both as CSV files.
4. Everything so far was only with post-training quantization. QAT is still an open question. Without QAT, we found that int8 alone on the ResNet was insufficient, however first layer int16 and remaining layers int8 was found to work quite well. It is unclear if QAT will fix this, though it is on our list of things to investigate. Quantization and supporting of the transformer section on NSS is ongoing, currently everything there is running on bf16 precision.
5. Here are two csv files comparing the fp32 with our end-to-end pipline: 
[action_compare_full.csv](https://github.com/user-attachments/files/29012227/action_compare_full.csv)
[action_compare_summary.csv](https://github.com/user-attachments/files/29012226/action_compare_summary.csv)
6. Agreed. I'll try to run this pipeline on your model first, then try to validate our process on the community model.
7. Yes. I'm first going to see if I can push the quantization a bit further as well as validate your trained model, and ~~at week's end~~ _early next week (week of June 22)_ I'll share everything I can so that your team can verify the model conversion process and model performance on your end.

Thanks,
Nicholas

**Comment by nchristo-synaptics** [comment]
Our pipeline works for both your trained model and the community model -- both run in <700ms / inference. I've attached a short summary for each below.

[aloha_condensed.md](https://github.com/user-attachments/files/29068688/aloha_condensed.md)
[quan1328_condensed.md](https://github.com/user-attachments/files/29068689/quan1328_condensed.md)

Best,
Nicholas

**Comment by nchristo-synaptics** [comment]
Hi @quanguyen and @vietnv,

Quick update: we now have a mixed-precision version of the model running on the NSS at ~440 ms per inference with reasonable accuracy.

Going to full int8 quantization, by contrast, causes a significant accuracy drop. It may be possible to recover some of that with quantization-aware training (QAT) — I'm happy to investigate that as a follow-up if it would be valuable.

At this point we're getting close to the practical runtime limit of the NSS for this model. ~440 ms corresponds to roughly 2–3 fps, so before investing further it would help to align on your target:

- Is 2–3 fps acceptable for your use case? If not, what frame rate are you aiming for?

One thing worth flagging up front: 10 fps is likely not achievable on this hardware. Even a fully int8-quantized ResNet backbone on its own runs at ~120 ms per inference — and that's before the transformer, which is actually the larger part of the model.

Happy to discuss the trade-offs.

**Comment by nchristo-synaptics** [comment]
Current status is ~320ms / inference in bf16 precision for a 2xQVGA input ACT model with standard lerobot parameters, and ~200ms for the same model if quantized under the mixed-precision quantization scheme discussed above.
